### PR TITLE
Add nearest neighbor blueprint and create it in attribute blueprint f…

### PR DESCRIPTION
**Pull request by geirst**
…actory.

@havardpe please review
@arnej27959 FYI

### CODE REVIEW

File: searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp [hunk 1]
```diff
@@ -210,5 +228,96 @@ TEST("requireThatFastSearchLocationTermsWork") {
...
+    Config cfg(BasicType::INT32, CollectionType::SINGLE);
+    return AttributeFactory::createAttribute(name, cfg);
+}
+
+template <typename BlueprintType>
+const BlueprintType*
+as_type(const Blueprint& blueprint)
+{
+    const auto* result = dynamic_cast<const BlueprintType*>(&blueprint);
+    ASSERT_TRUE(result != nullptr);
```
Comment (havardpe) [inline searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp]: if you assert here, why not just return by reference?

File: searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp [hunk 2]
```diff
@@ -210,5 +228,96 @@ TEST("requireThatFastSearchLocationTermsWork") {
...
+
+void
+expect_empty_blueprint(AttributeVector::SP attr, const TensorSpec& query_tensor, bool insert_query_tensor = true)
+{
+    NearestNeighborFixture f(attr);
+    if (insert_query_tensor) {
+        f.set_query_tensor(query_tensor);
+    }
+    auto result = f.create_blueprint();
+    as_type<EmptyBlueprint>(*result);
```
Comment (havardpe) [inline searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp]: this testing is very indirect and confusing. If this is a use-case, consider doing just casting in the as_type function and asserting/expecting the result outside

File: searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp [hunk 3]
```diff
@@ -210,5 +228,96 @@ TEST("requireThatFastSearchLocationTermsWork") {
...
+TEST("empty blueprint is created when nearest neighbor term is invalid")
+{
+    TensorSpec sparse_x = TensorSpec("tensor(x{})").add({{"x", 0}}, 3);
+    TensorSpec dense_y_2 = TensorSpec("tensor(y[2])").add({{"y", 0}}, 3).add({{"y", 1}}, 5);
+    expect_empty_blueprint(make_int_attribute(field)); // attribute is not a tensor
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x{})")); // attribute is not a dense tensor
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x[2],y[2])")); // tensor type is not of order 1
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x[2])")); // query tensor not found
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x[2])"), sparse_x); // query tensor is not dense
+    expect_empty_blueprint(make_tensor_attribute(field, "tensor(x[2])"), dense_y_2); // tensor types are not equal
```
Comment (havardpe) [inline searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp]: should also test x[2] vs x[3] (not same size)
and float x[2] vs double x[2] -> ok

File: searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp [hunk 4]
```diff
@@ -592,11 +602,43 @@ class CreateBlueprintVisitor : public CreateBlueprintVisitorHelper
...
+        if (dense_query_tensor == nullptr) {
+            return fail_nearest_neighbor_term(n, make_string("Query tensor is not a dense tensor (type=%s)",
+                                                             query_tensor->type().to_spec().c_str()));
+        }
+        if (dense_attr_tensor->getTensorType() != dense_query_tensor->type()) {
+            // TODO: consider allowing different data types (float vs double).
+            return fail_nearest_neighbor_term(n, make_string("Attribute tensor type (%s) and query tensor type (%s) are not equal",
+                                                             dense_attr_tensor->getTensorType().to_spec().c_str(), dense_query_tensor->type().to_spec().c_str()));
+        }
+        query_tensor.release();
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp]: this makes the code less exception safe than it needs to be

File: searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp [hunk 5]
```diff
@@ -592,11 +602,43 @@ class CreateBlueprintVisitor : public CreateBlueprintVisitorHelper
...
+        if (query_tensor.get() == nullptr) {
+            return fail_nearest_neighbor_term(n, "Query tensor was not found in request context");
+        }
+        auto* dense_query_tensor = dynamic_cast<DenseTensorView*>(query_tensor.get());
+        if (dense_query_tensor == nullptr) {
+            return fail_nearest_neighbor_term(n, make_string("Query tensor is not a dense tensor (type=%s)",
+                                                             query_tensor->type().to_spec().c_str()));
+        }
+        if (dense_attr_tensor->getTensorType() != dense_query_tensor->type()) {
+            // TODO: consider allowing different data types (float vs double).
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp]: comparing type().dimensions() here will allow for float/double cells

**Review by havardpe — COMMENTED** [review]
please consider my comments

File: searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp (hunk 5 above)
Comment (geirst) [reply searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp]: Thanks! Talked with @arnej27959, and I think we will defer adding support for this until we have something working end to end.

File: searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp (hunk 1 above)
Comment (geirst) [reply searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp]: Fixed

File: searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp (hunk 2 above)
Comment (geirst) [reply searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp]: Fixed

File: searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp (hunk 3 above)
Comment (geirst) [reply searchlib/src/tests/attribute/searchable/attributeblueprint_test.cpp]: Fixed

File: searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp (hunk 4 above)
Comment (geirst) [reply searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp]: Fixed

**Comment by geirst** [comment]
@havardpe PTAL

**Review by havardpe — APPROVED** [review]

**Pull request merged**